### ISSUE TITLE
instance segmentation exporting model to saved model

### ISSUE THREAD

**Issue by Danzip** [ultralytics/ultralytics#2730]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi So my question/issue is what to make of the outputs im getting from my converted keras model. 
while ultralytics gives a very clear its unclear what im getting with keras. 
Ive added sampled code:
```
import os
from matplotlib import pyplot as plt
import cv2
import numpy as np
from ultralytics import YOLO
import tensorflow as tf
# input img

bus = "bus.jpg"

# ultra model

model = YOLO('yolov8n-seg.pt')

# create keras

model.export(format="saved_model")

# load keras

keras_model = tf.keras.models.load_model('yolov8n-seg_saved_model')

# ultra inference

results = model(bus)  # predict on an image
res_plotted = results[0].plot()
plt.imshow(res_plotted)
plt.show()


# keras inference

def preprocess(image, target_size):
    '''

    Args:
        image: original image
        target_size: YOLO input size

    Returns:
        resized image matching YOLO input size
    '''
    ih, iw = target_size, target_size
    h, w, _ = image.shape
    scale = min(iw / w, ih / h)
    nw, nh = int(scale * w), int(scale * h)
    image_resized = cv2.resize(image, (nw, nh))
    image_padded = np.full(shape=[ih, iw, 3], fill_value=128.0)
    dw, dh = (iw - nw) // 2, (ih - nh) // 2
    image_padded[dh:nh + dh, dw:nw + dw, :] = image_resized
    image_padded = image_padded / 255.
    image_padded = tf.convert_to_tensor(image_padded, dtype=tf.float32)
    image_padded = tf.expand_dims(image_padded, 0)
    return image_padded
# keras inference
bus_img = cv2.imread(bus)
bus_img = preprocess(bus_img, 640)
results_keras = keras_model(bus_img

)
print(results_keras)
```
![bus](https://github.com/ultralytics/ultralytics/assets/34044636/25504044-a0d7-4eb1-ad30-aafd678e8ed2)
![image](https://github.com/ultralytics/ultralytics/assets/34044636/e28e0f7e-787f-4ef6-9f6e-7a3343b87ce8)
![image](https://github.com/ultralytics/ultralytics/assets/34044636/624248b9-0368-4313-8c11-ffc39004df65)

can someone please explain how to use the data given from the keras model?

what is the tensor (1,116,8400), and the mask (1,160,160,32) why is it 32? shouldn't it be 80 like the amount of classes?
did I export correctly?

**Comment by glenn-jocher** [comment]
Hi @Danzip,

To answer your questions, the `(1,116,8400)` tensor output from the Keras model represents the flattened segmentation head output tensor after training. In general, the memory usage of an instance segmentation network depends on its resolution and the number of objects it is intended to detect.

Regarding your question on the output tensor size, the number 32 corresponds to the number of segmentation classes as defined during training. This is different from the number of object detection classes, which is usually 1 less than the total number of classes in an instance segmentation model, so this seems correct.

Regarding the export, when you use the `model.export(format="saved_model")` function, it saves the model in TensorFlow's SavedModel format that TensorFlow Serving uses to serve models. The format should be correct, although a closer examination of your model and the export process may be required to ensure that everything is working correctly.

I hope this helps clarify things for you!

**Comment by Danzip** [comment]
Thank you so much  for answering my question so quickly!

**1.** I'm still struggling to understand the meaning of the tensors does ```(1,116,8400)``` mean batch of 1, 
116 proposals and an output mask that flattened has 8400 params in it? 
**2.** as you can see from my example I'm loading the yolov8n-seg pretrained without any changes. 
Is it trained on ms coco?
**3.** if its not to much to ask can you tell me what does 8400 mean? how do I extract the 2d mask from it?

**4.** also regarding the other tensor ```(1,160,160,32)```, 
My intuition was that the ```(1,160,160,32)```, was the mask becuase the input size is 640 and we get a quarter of it at the output. 
**5**.I'm still confused by the num of classes(32) can you clarify for me why 32 classes make sense with coco i thought it had 80 classes.

I'm glad that the export looks good to you, It means I'm very close to a solution.
Thank you again for your help

**Comment by glenn-jocher** [comment]
@Danzip you're welcome, happy to help! 

1. Yes, you are correct that `(1,116,8400)` means a batch of 1 with 116 proposals and flattened output mask with 8400 parameters.

2. Yes, the `yolov8n-seg` pretrained model was trained on MS COCO.

3. The `8400` represents the flattened output mask of the model. To view the 2D mask, you will need to reshape the tensor from `(1, 116, 8400)` to `(1, 116, h, w, c)`, where `h` and `w` are the height and width of the input image and `c` is the number of segmentation classes. You can then access the 2D mask at each proposal as a numpy array like `mask = tensor[0][proposal_index].numpy().reshape(h, w, c)`.

4. You are correct that the `(1,160,160,32)` tensor represents the mask output of the model. It is a 2D mask of size `(160, 160)` for each of the 32 classes in the model.

5. The number of classes in the output mask tensor is based on the number of segmentation classes in the model, which in this case is 32. It is different from the number of object detection classes which is generally one less than the total number of classes in an instance segmentation model. While MS COCO has 80 classes, the YOLOv8-S has been designed to use only 32 segmentation classes for this task.

Great to hear that the export looks good. I hope this helps clear things up for you!

**Comment by lvj5077** [comment]
@glenn-jocher Hi, I am also struggling to understand why 32 classes. I am working with mlmodel, which also gives me two arrays as output one is (1, 116, 8400), and another is (1,160,160,32). However, I can detect objects like "books", and "clocks" which are very indexed really behind on coco.yaml. If there are only 32 classes (be trained) in the model, what are they? Or the "32" indicates the top 32 classes the model detects?

**Comment by glenn-jocher** [comment]
@lvj5077 hello! The number 32 corresponds to the number of segmentation classes in the model, which is different from the number of object detection classes. While MS COCO has 80 object detection classes, YOLOv8 is designed to use only 32 segmentation classes for this task. These classes include things like 'person', 'car', 'bus', 'dog', 'cat', etc. However, it is possible for the model to detect objects that are not included in these segmentation classes, but it might not be able to provide accurate segmentation for them. I hope this helps clarify things for you!
